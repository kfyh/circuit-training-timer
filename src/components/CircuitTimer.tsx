import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { StartStopButton, TimeLabel, ExerciseLabel } from './timer';
import { StepIntervalController, SetIntervalController } from '../controllers';

type CircuitTimerProps = {
	intervals: Array<object>;
};

type CircuitTimerState = {
	controller: any;
	currentIntervalIndex: number;
	currentIntervalRep: number;
	currentStepIndex: number;
	currentStepRep: number;
	isRunning: boolean;
	endTime: number;
	timeLeft: number;
}
export class CircuitTimer extends React.Component<CircuitTimerProps, CircuitTimerState> {
	private timer = NaN;

	constructor(props: CircuitTimerProps) {
		super(props);

		this.state = {
			controller: null,
			currentIntervalIndex: 0,
			currentIntervalRep: 1,
			currentStepIndex: 0,
			currentStepRep: 1,
			isRunning: false,
			endTime: 0,
			timeLeft: 0,
		};
	}

	public componentWillUnmount(): void {
		this.stopTimer();
	}

	public render(): ReactElement {
		let currentInterval = this.props.intervals[this.state.currentIntervalIndex];
		const currentRep = this.state.currentIntervalRep;
		if (currentInterval.type === 'set') {
			currentInterval = currentInterval.steps[this.state.currentStepIndex];
		}

		return (
			<div>
				<ExerciseLabel interval={currentInterval} currentRep={currentRep} />
				<TimeLabel time={this.state.timeLeft} />
				<StartStopButton isRunning={this.state.isRunning} onClick={this.handleClick} />
			</div>
		);
	}

	private handleClick = (): void => {
		if (this.state.isRunning) {
			this.stopTimer();
		} else {
			this.startTimer();
		}
	};

	private startTimer = (): void => {
		if (this.state.isRunning) {
			return;
		}
		let controller = this.state.controller;
		let timeLeft = this.state.timeLeft;
		if (controller === null) {
			const interval = this.props.intervals[this.state.currentIntervalIndex];
			if (interval.type === 'set') {
				controller = new SetIntervalController(interval);
			} else {
				controller = new StepIntervalController(interval);
			}
			controller.start(Date.now());
			timeLeft = controller.timeLeft;
		} else {
			controller.resume(Date.now());
			timeLeft = controller.timeLeft;
		}

		this.setState({
			controller: controller,
			isRunning: true,
			endTime: Date.now() + timeLeft,
		});

		this.timer = setInterval(this.update, 100);
	};

	private stopTimer = (): void => {
		if (this.state.isRunning) {
			this.setState({
				isRunning: false,
			});
			this.state.controller.pause();
			clearInterval(this.timer);
		}
	};

	private update = (): void => {
		const values = this.state.controller.update(Date.now());
		if (values.isComplete) {
			const currentIntervalIndex = this.state.currentIntervalIndex + 1;
			if (currentIntervalIndex >= this.props.intervals.length) {
				this.endTimer();
			} else {
				this.setNextInterval(currentIntervalIndex);
			}
		} else {
			this.setState({
				timeLeft: values.timeLeft,
				currentIntervalRep: values.reps,
				currentStepIndex: values.currentStepIndex ? values.currentStepIndex : 0,
			});
		}
	};

	private endTimer = (): void => {
		clearInterval(this.timer);
		this.setState({
			controller: null,
			currentIntervalIndex: 0,
			currentIntervalRep: 1,
			currentStepIndex: 0,
			currentStepRep: 1,
			isRunning: false,
			endTime: 0,
			timeLeft: 0,
		});
		this.props.history.push('/');
	};

	setNextInterval = (interval) => {
		const nextInterval = this.props.intervals[interval];
		let nextController;
		if (nextInterval.type === 'set') {
			nextController = new SetIntervalController(nextInterval);
		} else {
			nextController = new StepIntervalController(nextInterval);
		}
		nextController.start(Date.now());
		this.setState({
			controller: nextController,
			timeLeft: nextController.timeLeft,
			currentIntervalIndex: interval,
			currentIntervalRep: nextController.reps,
			endTime: nextController.endTime,
		});
	};
}

const mapStateToProps = (state) => {
	return {
		intervals: state.currentCircuit,
	};
};

export const ConnectedCircuitTimer = connect(mapStateToProps)(CircuitTimer);
