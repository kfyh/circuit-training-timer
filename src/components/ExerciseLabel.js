import React from 'react'
import PropTypes from 'prop-types'

function ExerciseLabel (props) {
  const interval = props.interval
  const excerciseLabel = interval.label
  const totalReps = interval.reps

  return (
    <div>{excerciseLabel} ({props.currentRep}/{totalReps})</div>
  )
}

ExerciseLabel.propTypes = {
  interval: PropTypes.object,
  currentRep: PropTypes.number
}

export default ExerciseLabel
