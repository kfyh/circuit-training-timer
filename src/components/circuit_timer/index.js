export * from './StartStopButton.js'
export * from './ExerciseLabel.js'
export * from './TimeLabel.js'
