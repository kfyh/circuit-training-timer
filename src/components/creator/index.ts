export * from './AddExercisePage';
export * from './ExerciseForm';