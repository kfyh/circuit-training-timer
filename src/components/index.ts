export * from './CircuitTimer';
export * from './selector/CircuitListView';
export * from './timer/index';
export * from './Dashboard';
export * from './creator/index';
