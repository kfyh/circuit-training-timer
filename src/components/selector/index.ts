export * from './CircuitListView';
export * from './CircuitView';
