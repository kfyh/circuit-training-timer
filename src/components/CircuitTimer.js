import React from 'react'
import PropTypes from 'prop-types'
import { StartStopButton, TimeLabel, ExerciseLabel } from './circuit_timer'
import { StepIntervalController, SetIntervalController } from '../controllers'

class CircuitTimer extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      controller: null,
      currentIntervalIndex: 0,
      currentIntervalRep: 1,
      currentStepIndex: 0,
      currentStepRep: 1,
      isRunning: false,
      endTime: 0,
      timeLeft: 0
    }
  }

  componentDidMount () {
  }

  componentWillUnmount () {
    this.stopTimer()
  }

  render () {
    let currentInterval = this.props.intervals[this.state.currentIntervalIndex]
    let currentRep = this.state.currentIntervalRep
    if (currentInterval.type === 'set') {
      currentInterval = currentInterval.steps[this.state.currentStepIndex]
      currentRep = this.state.currentStepRep
    }

    return (
      <div>
        <ExerciseLabel interval={currentInterval} currentRep={currentRep} />
        <TimeLabel time={this.state.timeLeft}/>
        <StartStopButton isRunning={this.state.isRunning} onClick={this.handleClick} />
      </div>
    )
  }

  handleClick = () => {
    if (this.state.isRunning) {
      this.stopTimer()
    } else {
      this.startTimer()
    }
  }

  startTimer = () => {
    if (this.state.isRunning) {
      return
    }
    let controller = this.state.controller
    let timeLeft = this.state.timeLeft
    if (controller === null) {
      const interval = this.props.intervals[this.state.currentIntervalIndex]
      if (interval.type === 'set') {
        controller = new SetIntervalController(interval)
      } else {
        controller = new StepIntervalController(interval)
      }
      controller.start(Date.now())
      timeLeft = controller.timeLeft
    } else {
      controller.resume(Date.now())
      timeLeft = controller.timeLeft
    }

    this.setState({
      controller: controller,
      isRunning: true,
      endTime: Date.now() + timeLeft
    })

    this.timer = setInterval(this.update, 100)
  }

  stopTimer = () => {
    if (this.state.isRunning) {
      this.setState({
        isRunning: false
      })
      this.state.controller.pause()
      clearInterval(this.timer)
    }
  }

  update = () => {
    const values = this.state.controller.update(Date.now())
    if (values.isComplete) {
      const currentIntervalIndex = this.state.currentIntervalIndex + 1
      if (currentIntervalIndex >= this.props.intervals.length) {
        clearInterval(this.timer)
        this.setState({
          controller: null,
          currentIntervalIndex: 0,
          currentIntervalRep: 1,
          currentStepIndex: 0,
          currentStepRep: 1,
          isRunning: false,
          endTime: 0,
          timeLeft: 0
        })
        this.props.onComplete()
      } else {
        const nextInterval = this.props.intervals[currentIntervalIndex]
        let nextController
        if (nextInterval.type === 'set') {
          nextController = new SetIntervalController(nextInterval)
        } else {
          nextController = new StepIntervalController(nextInterval)
        }
        nextController.start(Date.now())
        this.setState({
          controller: nextController,
          timeLeft: nextController.timeLeft,
          currentIntervalIndex: currentIntervalIndex,
          currentIntervalRep: nextController.reps,
          endTime: nextController.endTime
        })
      }
    } else {
      this.setState({
        timeLeft: values.timeLeft,
        currentIntervalRep: values.reps,
        currentStepIndex: values.currentStepIndex ? values.currentStepIndex : 0
      })
    }
  }
}

CircuitTimer.propTypes = {
  intervals: PropTypes.arrayOf(PropTypes.object),
  onComplete: PropTypes.func
}

export default CircuitTimer
