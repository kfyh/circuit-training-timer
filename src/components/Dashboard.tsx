import React, { ReactElement } from 'react';
import { connect } from 'react-redux';
import { ISelectorAction, ISelectorReducerState } from '../reducers';
import { Circuit } from '../types/circuits';
import { Dispatch } from 'redux';
import { selectCircuit } from '../actions/selectorActions';
import { CircuitListView } from './selector/';
import { History } from 'history';
import { NavLink } from 'react-router-dom';

type DashboardProps = {
	circuits: Array<Circuit>;
	history: History;
	selectCircuit: (circuit: Circuit) => void;
};

export class Dashboard extends React.Component<DashboardProps, Record<string, never>> {
	constructor(props: DashboardProps) {
		super(props);
	}

	private onClick = (circuit: Circuit) => {
		this.props.selectCircuit(circuit);
		this.props.history.push('/timer');
	};

	public render(): ReactElement {
		return (
			<div>
				<p>Dashboard</p>
				<a>New Circuit</a>
				<NavLink to="/addexercise" exact={true}>
					Add Exercise
				</NavLink>
				<CircuitListView circuits={this.props.circuits} onSelect={this.onClick} />
			</div>
		);
	}
}

const mapStateToProps = (state: ISelectorReducerState) => {
	return {
		circuits: state.circuits,
	};
};

const mapDispatchToProps = (dispatch: Dispatch<ISelectorAction>) => ({
	selectCircuit: (circuit: Circuit) => dispatch(selectCircuit(circuit)),
});

export const ConnectedDashboard = connect(mapStateToProps, mapDispatchToProps)(Dashboard);
