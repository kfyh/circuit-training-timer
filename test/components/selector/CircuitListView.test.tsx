import React from 'react';
import { shallow } from 'enzyme';
import { CircuitListView } from '../../../src/components/selector/';
import { circuit } from '../../fixtures/exerciseFixtures';

test('Given CircuitListView, it renders correctly', () => {
	const render = shallow(<CircuitListView circuits={[circuit]} onSelect={jest.fn()} />);
	expect(render).toMatchSnapshot();
});
