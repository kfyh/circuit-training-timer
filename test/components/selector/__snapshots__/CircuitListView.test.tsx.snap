// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given CircuitListView, it renders correctly 1`] = `
<div>
  <CircuitView
    circuit={
      Object {
        "exerciseGroups": Array [],
        "id": "0",
        "name": "circuit",
        "repetitions": 1,
      }
    }
    key="0"
    onClick={[MockFunction]}
  />
</div>
`;
